File: Java/src/main/java/QuestionBank/Solution101_200/Solution112.java
package QuestionBank.Solution101_200;

import tool.TreeNode;

/**
 * @author windows
 */
public class Solution112 {
    public static void main(String[] args) {
        Solution112 solution112 = new Solution112();
    }

    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) {
            return false;
        }
        if (root.left == null && root.right == null) {
            return targetSum - root.val == 0;
        }
        boolean left = hasPathSum(root.left, targetSum - root.val);
        boolean right = hasPathSum(root.right, targetSum - root.val);
        return left || right;
    }
}

File: C++/src/Solution101_200/Solution112.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution112.h"

bool Solution112::hasPathSum(TreeNode *root, int targetSum) {
    return dfs(root, targetSum, 0);
}


bool Solution112::dfs(TreeNode *root, int targetSum, int sum) {
    if (root == nullptr) return false;
    if (root-> left == nullptr && root->right == nullptr){
        return sum + root->val == targetSum;
    }
    bool left = dfs(root->left, targetSum, sum + root->val);
    bool right = dfs(root->right, targetSum, sum + root->val);
    return left || right;
}